org.eclipse.lsp4e/src/org/eclipse/lsp4e/operations/typeHierarchy/TypeHierarchyContentProvider.java

```diff
@@ -0,0 +1,77 @@
+/*******************************************************************************
+ * Copyright (c) 2023 Red Hat Inc. and others.
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *******************************************************************************/
+package org.eclipse.lsp4e.operations.typeHierarchy;
+
+import java.util.concurrent.TimeUnit;
+
+import org.eclipse.jface.text.IDocument;
+import org.eclipse.jface.text.ITextSelection;
+import org.eclipse.jface.viewers.ITreeContentProvider;
+import org.eclipse.lsp4e.LSPEclipseUtils;
+import org.eclipse.lsp4e.LanguageServerPlugin;
+import org.eclipse.lsp4e.LanguageServers;
+import org.eclipse.lsp4e.LanguageServersRegistry.LanguageServerDefinition;
+import org.eclipse.lsp4j.Position;
+import org.eclipse.lsp4j.TypeHierarchyItem;
+import org.eclipse.lsp4j.TypeHierarchyPrepareParams;
+import org.eclipse.lsp4j.TypeHierarchySubtypesParams;
+
+public class TypeHierarchyContentProvider implements ITreeContentProvider {
+
+	private final LanguageServerDefinition lsDefinition;
+	private IDocument document;
+
+	public TypeHierarchyContentProvider(LanguageServerDefinition lsDefinition, IDocument document) {
+		this.lsDefinition = lsDefinition;
+		this.document = document;
+	}
+
+	@Override
+	public Object[] getElements(Object inputElement) {
+		if (inputElement instanceof ITextSelection textSelection) {
+			try {
+				Position position = LSPEclipseUtils.toPosition(textSelection.getOffset(), document);
+				TypeHierarchyPrepareParams prepare = new TypeHierarchyPrepareParams(LSPEclipseUtils.toTextDocumentIdentifier(document), position);
+				LanguageServers.forDocument(document).withPreferredServer(lsDefinition)
+					.computeFirst(ls -> ls.getTextDocumentService().prepareTypeHierarchy(prepare))
+					.thenApply(res -> res.map(list -> list.toArray(Object[]::new)).orElse(new Object[0]))
+					.get(500, TimeUnit.MILLISECONDS);
+			} catch (Exception e) {
+				LanguageServerPlugin.logError(e);
+			}
+		}
+		return new Object[0];
+	}
+
+	@Override
+	public Object[] getChildren(Object parentElement) {
+		if (parentElement instanceof TypeHierarchyItem parentItem) {
+			try {
+				LanguageServers.forDocument(document).withPreferredServer(lsDefinition)
```

Comment:
I think you must be sure to target the LS that answered the prepareTypeHierarchy, so in the method above you should remember the wrapper, and use `wrapper.execute(ls -> ls....)` in here. You should find a couple of example in the code base, for example for prepare rename & rename.

Reply:
Hi @mickaelistria , do you plan to follow on this comment? The code might fail like this if there is more than one server that has the required capability.

Reply:
I don't think I'll work on it until I face that issue in one of the integrations I'm dealing with. So if you feel interested in implementing it, that would be very welcome!